hack/boilerplate/boilerplate.py

```diff
@@ -0,0 +1,242 @@
+    return p.search(data)
+
+
+def file_passes(filename, refs, regexs):
+    try:
+        f = open(filename, 'r')
+    except Exception as exc:
+        print("Unable to open %s: %s" % (filename, exc), file=verbose_out)
+        return False
+
+    data = f.read()
+    f.close()
+
+    # determine if the file is automatically generated
+    generated = is_generated_file(filename, data, regexs)
+
+    basename = os.path.basename(filename)
+    extension = file_extension(filename)
+    if generated:
+        if extension == "go":
+            extension = "generatego"
+        elif extension == "bzl":
+            extension = "generatebzl"
+
+    if extension != "":
+        ref = refs[extension]
+    else:
+        ref = refs[basename]
+
+    # remove extra content from the top of files
+    if extension == "go" or extension == "generatego":
+        p = regexs["go_build_constraints"]
+        (data, found) = p.subn("", data, 1)
+    elif extension in ["sh", "py"]:
+        p = regexs["shebang"]
+        (data, found) = p.subn("", data, 1)
+
+    data = data.splitlines()
+
+    # if our test file is smaller than the reference it surely fails!
+    if len(ref) > len(data):
+        print('File %s smaller than reference (%d < %d)' %
+              (filename, len(data), len(ref)),
+              file=verbose_out)
+        return False
+
+    # trim our file to the same number of lines as the reference file
+    data = data[:len(ref)]
+
+    p = regexs["year"]
+    for d in data:
+        if p.search(d):
+            if generated:
+                print('File %s has the YEAR field, but it should not be in generated file' %
+                      filename, file=verbose_out)
+            else:
+                print('File %s has the YEAR field, but missing the year of date' %
+                      filename, file=verbose_out)
+            return False
+
+    if not generated:
+        # Replace all occurrences of the regex "2014|2015|2016|2017|2018" with "YEAR"
+        p = regexs["date"]
+        for i, d in enumerate(data):
+            (data[i], found) = p.subn('YEAR', d)
+            if found != 0:
+                break
+
+    # if we don't match the reference at this point, fail
+    if ref != data:
+        print("Header in %s does not match reference, diff:" %
+              filename, file=verbose_out)
+        if args.verbose:
+            print(file=verbose_out)
+            for line in difflib.unified_diff(ref, data, 'reference', filename, lineterm=''):
+                print(line, file=verbose_out)
+            print(file=verbose_out)
+        return False
+
+    return True
+
+
+def file_extension(filename):
+    return os.path.splitext(filename)[1].split(".")[-1].lower()
+
+
+skipped_dirs = ['Godeps', 'third_party', '_gopath', '_output', '.git', 'cluster/env.sh',
+                "vendor", "test/e2e/generated/bindata.go", "hack/boilerplate/test",
+                "staging/src/k8s.io/kubectl/pkg/generated/bindata.go"]
```

Comment:
do we have this directory?

Reply:
these directories are hardcoded in this k8s script, we don't have these directories, it's ok to remove or keep them

Reply:
fixed in e095abbe072cfdd720c6c1ab560c352cd59e053c, hardcoded rules are ignored (except the script itself)